File: frontend/src/services/dca.ts
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || '';

export interface DCABot {
  id: string;
  symbol: string;
  investment_amount: number;
  interval_minutes: number;
  take_profit_percent: number;
  safety_order_multiplier: number;
  max_safety_orders: number;
  price_deviation_percent: number;
  status: string;
  total_invested: number;
  total_sold: number;
  current_position_qty: number;
  avg_entry_price: number;
  created_at: string;
  started_at: string | null;
  stopped_at: string | null;
}

export interface DCAOrder {
  id: string;
  bot_id: string;
  order_type: string; // BASE, SAFETY, TAKE_PROFIT
  side: string;
  quantity: number;
  price: number;
  total: number;
  status: string;
  order_number: number;
  executed_at: string | null;
  created_at: string;
}

export interface DCACreateRequest {
  symbol: string;
  investment_amount: number;
  interval_minutes: number;
  take_profit_percent: number;
  safety_order_multiplier: number;
  max_safety_orders: number;
  price_deviation_percent: number;
}

function getAuthToken() {
  if (typeof window === 'undefined') return '';
  return localStorage.getItem('auth_token') || '';
}

export const dcaApi = {
  async createBot(bot: DCACreateRequest): Promise<DCABot> {
    const res = await fetch(`${API_BASE_URL}/api/dca/bots`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${getAuthToken()}` },
      body: JSON.stringify(bot),
    });
    if (!res.ok) throw new Error('Failed to create DCA bot');
    return res.json();
  },

  async getBots(): Promise<DCABot[]> {
    const res = await fetch(`${API_BASE_URL}/api/dca/bots`, {
      headers: { Authorization: `Bearer ${getAuthToken()}` },
    });
    if (!res.ok) throw new Error('Failed to get DCA bots');
    return res.json();
  },

  async startBot(id: string): Promise<void> {
    const res = await fetch(`${API_BASE_URL}/api/dca/bots/${id}/start`, {
      method: 'POST',
      headers: { Authorization: `Bearer ${getAuthToken()}` },
    });
    if (!res.ok) throw new Error('Failed to start DCA bot');
  },

  async stopBot(id: string): Promise<void> {
    const res = await fetch(`${API_BASE_URL}/api/dca/bots/${id}/stop`, {
      method: 'POST',
      headers: { Authorization: `Bearer ${getAuthToken()}` },
    });
    if (!res.ok) throw new Error('Failed to stop DCA bot');
  },

  async deleteBot(id: string): Promise<void> {
    const res = await fetch(`${API_BASE_URL}/api/dca/bots/${id}`, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${getAuthToken()}` },
    });
    if (!res.ok) throw new Error('Failed to delete DCA bot');
  },

  async getBotOrders(id: string): Promise<DCAOrder[]> {
    const res = await fetch(`${API_BASE_URL}/api/dca/bots/${id}/orders`, {
      headers: { Authorization: `Bearer ${getAuthToken()}` },
    });
    if (!res.ok) throw new Error('Failed to get bot orders');
    return res.json();
  },
};

File: backend/internal/domain/model/dca.go
package model

import "time"

// DCABotStatus constants
const (
	DCABotStatusStopped   = "STOPPED"
	DCABotStatusRunning   = "RUNNING"
	DCABotStatusCompleted = "COMPLETED"
	DCABotStatusError     = "ERROR"
)

// DCAOrderType constants
const (
	DCAOrderTypeBase       = "BASE"
	DCAOrderTypeSafety     = "SAFETY"
	DCAOrderTypeTakeProfit = "TAKE_PROFIT"
)

// DCABot represents a Dollar Cost Averaging bot configuration
type DCABot struct {
	ID                    string     `json:"id"`
	UserID                string     `json:"user_id"`
	Symbol                string     `json:"symbol"`
	InvestmentAmount      float64    `json:"investment_amount"`
	IntervalMinutes       int        `json:"interval_minutes"`
	TakeProfitPercent     float64    `json:"take_profit_percent"`
	SafetyOrderMultiplier float64    `json:"safety_order_multiplier"`
	MaxSafetyOrders       int        `json:"max_safety_orders"`
	PriceDeviationPercent float64    `json:"price_deviation_percent"`
	Status                string     `json:"status"`
	TotalInvested         float64    `json:"total_invested"`
	TotalSold             float64    `json:"total_sold"`
	CurrentPositionQty    float64    `json:"current_position_qty"`
	AvgEntryPrice         float64    `json:"avg_entry_price"`
	CreatedAt             time.Time  `json:"created_at"`
	UpdatedAt             time.Time  `json:"updated_at"`
	StartedAt             *time.Time `json:"started_at,omitempty"`
	StoppedAt             *time.Time `json:"stopped_at,omitempty"`
}

// DCAOrder represents a DCA order (base, safety, or take-profit)
type DCAOrder struct {
	ID           string     `json:"id"`
	BotID        string     `json:"bot_id"`
	OrderType    string     `json:"order_type"`
	Side         string     `json:"side"`
	Quantity     float64    `json:"quantity"`
	Price        float64    `json:"price"`
	Total        float64    `json:"total"`
	Status       string     `json:"status"`
	OrderNumber  int        `json:"order_number"`
	ExecutedAt   *time.Time `json:"executed_at,omitempty"`
	CreatedAt    time.Time  `json:"created_at"`
}

// DCABotCreateRequest represents the request to create a new DCA bot
type DCABotCreateRequest struct {
	Symbol                string  `json:"symbol"`
	InvestmentAmount      float64 `json:"investment_amount"`
	IntervalMinutes       int     `json:"interval_minutes"`
	TakeProfitPercent     float64 `json:"take_profit_percent"`
	SafetyOrderMultiplier float64 `json:"safety_order_multiplier"`
	MaxSafetyOrders       int     `json:"max_safety_orders"`
	PriceDeviationPercent float64 `json:"price_deviation_percent"`
}
